### PR TITLE
Conservation fix

### PR DESCRIPTION
**Pull request by LenkaNovak**
# PULL REQUEST

## Purpose and Content
Fixes non-conservation introduced by masking PR #114. Conservation due to radiation is also partially fixed, but small error still persists: see #120 

## Benefits and Risks
b: bug fix

## Linked Issues
(Provide references to any link issues. Use closes #issuenum to automatically close an open issue)
- Fixes (part of) #120

## PR Checklist
- [x] This PR has a corresponding issue OR is linked to an SDI.
- [x] I have followed CliMA's codebase [contribution](https://clima.github.io/ClimateMachine.jl/latest/Contributing/) and [style](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) guidelines OR N/A.
- [x] I have followed CliMA's [documentation policy](https://github.com/CliMA/policies/wiki/Documentation-Policy).
- [x] I have checked all issues and PRs and I certify that this PR does not duplicate an open PR.
- [x] I linted my code on my local machine prior to submission OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code used in an integration test OR N/A.
- [x] All tests ran successfully on my local machine OR N/A.
- [x] All classes, modules, and function contain docstrings OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: experiments/AMIP/moist_mpi_earth/slab_ocean/slab_init.jl [hunk 1]
```diff
@@ -37,11 +37,11 @@ function slab_ocean_space_init(::Type{FT}, space, p) where {FT}
...
 
 # ode
-function slab_ocean_rhs!(dY, Y, Ya, t)
-    p, F_aero, F_rad, land_mask = Ya
+function slab_ocean_rhs!(dY, Y, cache, t)
+    p, F_aero, F_rad, land_mask = cache
     FT = eltype(Y.T_sfc)
     rhs = @. -(F_aero + F_rad) / (p.h * p.ρ * p.c)
-    parent(dY.T_sfc) .= apply_mask.(FT, parent(land_mask), <, parent(rhs))
+    parent(dY.T_sfc) .= apply_mask.(FT, parent(land_mask), parent(rhs))
```
Comment (kmdeck) [inline experiments/AMIP/moist_mpi_earth/slab_ocean/slab_init.jl:44]: would it be clearer to use the ocean_mask here?

File: experiments/AMIP/moist_mpi_earth/slab_ice/slab_init.jl [hunk 2]
```diff
@@ -55,16 +55,16 @@ function ice_init(::Type{FT}; tspan, saveat, dt, space, ice_mask, stepper = Eule
...
-    problem = OrdinaryDiffEq.ODEProblem(ice_rhs!, Y, tspan, (; Ya = Ya, params = params))
+    problem = OrdinaryDiffEq.ODEProblem(ice_rhs!, Y, tspan, (; cache..., params = params))
     integrator = OrdinaryDiffEq.init(problem, stepper, dt = dt, saveat = saveat)
 
 
     SlabSimulation(params, Y, space, integrator)
 end
 
-get_ice_mask(h_ice, FT, threshold = 50) = (h_ice - FT(threshold)) > FT(0) ? FT(1) : FT(0)
+# get_ice_mask(h_ice, FT, threshold = 50) = (h_ice - FT(threshold)) > FT(0) ? FT(1) : FT(0)
```
Comment (kmdeck) [inline experiments/AMIP/moist_mpi_earth/slab_ice/slab_init.jl:67]: could remove if not needed in the future?

File: experiments/AMIP/moist_mpi_earth/slab_ocean/slab_init.jl (hunk 1 above)
Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/slab_ocean/slab_init.jl:44]: in the conservation checker, we multiply the e.g. ocean field by ocean_mask. Is it possible to do that here (do we still need `apply_mask`?

Comment (kmdeck) [reply experiments/AMIP/moist_mpi_earth/slab_ocean/slab_init.jl:44]: (similar question for the other slab RHS)

File: experiments/AMIP/moist_mpi_earth/coupler_utils/flux_calculator.jl [hunk 3]
```diff
@@ -70,15 +70,15 @@ function calculate_surface_fluxes_atmos_grid!(integrator, info_sfc)
     if :ρe_tot in propertynames(Y.c)
 
         flux_energy = ones(axes(dif_flux_energy))
-        parent(flux_energy) .= parent(tsf.shf .+ tsf.lhf .* swap_space!(abs.(ice_mask .- FT(1)), axes(tsf.shf)))  # only SHF above sea ice
+        parent(flux_energy) .= parent(tsf.shf .+ tsf.lhf .* swap_space!(FT(1) .- ice_mask, axes(tsf.shf)))  # only SHF above sea ice
```
Comment (kmdeck) [inline experiments/AMIP/moist_mpi_earth/coupler_utils/flux_calculator.jl:73]: maybe we could put a comment in about what 1-icemask indicates? I'm forgetting now :)

**Review by kmdeck — APPROVED** [review]
I think it looks good, Lenka! From my point of view, the main change is that each surface type now has its own mask, and these can be used directly, and then there were changes propagated to many files because of this.

File: experiments/AMIP/moist_mpi_earth/slab_ocean/slab_init.jl (hunk 1 above)
Comment (LenkaNovak) [reply experiments/AMIP/moist_mpi_earth/slab_ocean/slab_init.jl:44]: It's doing a slightly different thing here. In the conservation checker it's doing the weighted sum (so near the domain boundaries the ocean temperature is `T_sfc_ocean / ocean_fraction`), whereas here the `apply_mask` cuts out the ocean domain area, with `T_sfc_ocean` retaining the absolute value (i.e. not dividing by the `ocean_fraction`)

Comment (LenkaNovak) [reply experiments/AMIP/moist_mpi_earth/slab_ocean/slab_init.jl:44]: But I have changed the above to the ocean mask, which makes more sense. :)

**Comment by LenkaNovak** [comment]
I brought back the option to do non-monotonous remapping (for coarse enough resolution the monotonous map is too blocky), and changed the buildkite runs, so that both options are tested. Note that AMIP is much more sensitive to boundary conditions than slabplanet, so `h_elem` is set to 6, as in the ClimaAtmos default setting (which was modified after ClimaAtmos switched from Nq 5 to Nq 4).

**Pull request merged**